### ISSUE TITLE
Constraints on regular expression matches

### ISSUE THREAD

**Issue by rensink** [nl-utwente-groove/code#657]
If I use regular expressions on an edge, I often use it to find either the first or last element with some property in a list-like structure. \(For example, a rule to find the first occurence of a particular method call in a runtime stack.\) Regular expressions make it easy to find \*all\* elements with a certain property, but you need to duplicate parts of your rule in NAC-form to find only the first or last element, which can be quite distracting if you're trying to understand a rule for the first time.

For example, suppose I have a graph A-&gt;B-&gt;B-&gt;B-&gt;B-&gt;B-&gt;B-&gt;B
and some of the B nodes have a C node attached to them.

Now suppose that, starting from the A node, I wish to find all B-nodes with a C node attached. This is easy with a regular expression. \(See the attached grammar file\)

However, if I only want to find the first or last occurence of a B-node with a C attached, the rule starts to look a lot messier because you need to add a NAC stating that "No, there can't be another B with a C before/behind this occurence". And this NAC basically is a duplicate of what you're trying to find.

So my suggestion would be that you could specify in/around your regular expression "I only want the shortest/longest match of this regex". This should fix the NAC duplication problem, and I'm guessing this could also work faster performance-wise than trying to match a NAC. Also, in the edge case where there are multiple shortest or longest matches, I think it makes most sense to let them all count as separate matches, rather than randomly picking one.

For your info, this feature request purely is a small suggestion to make rewrite rules a bit more compact and easier to read. :\)

Kind regards,
Tim Molderez

\(PS: I also have a similar NAC duplication problem with rules that create something,but you only want them to match at most once. To implement this, you need to add a NAC that duplicates the exact same thing you're creating. .. but perhaps I should create a separate feature request for that..\)

Reported by: timmolderez

**Comment by rensink** [comment]
Grammar demonstrating the above example

Original comment by: timmolderez

**Comment by rensink** [comment]
- **priority**: 5 --> 4
- **assigned_to**: nobody --> maartenm42

Original comment by: rensink

**Comment by rensink** [comment]
The rest of the rule filters the matches for the regular expression, and only after that it is known which is the shortest/longest match. The regular expression therefore becomes context sensitive, and this undesirable. For this reason, it cannot be fixed this way.

Original comment by: maartenm42

**Comment by rensink** [comment]
- **status**: open --> closed

Original comment by: maartenm42

**Comment by rensink** [comment]
Just a small remark concerning you PS: the functionality you want already exists as a grammar property. Just set 'rhsIsNAC' to true to globally define implicit NACs to all RHSs of rules. 

Original comment by: zambon
